app/src/main/java/org/wikipedia/talk/NotificationDirectReplyHelper.kt

```diff
@@ -1,91 +1,75 @@
-                    )
-                }
+        CoroutineScope(Dispatchers.IO).launch(CoroutineExceptionHandler { _, throwable ->
+            L.e(throwable)
+            fallBackToTalkPage(context, title)
+        }) {
+            val token = async { CsrfTokenClient.getToken(wiki).blockingFirst() }
+            val talkPageResponse = async { ServiceFactory.getRest(wiki).getTalkPage(title.prefixedText) }
+            val topic = talkPageResponse.await().topics?.find {
+                it.id > 0 && it.html?.trim().orEmpty() == StringUtil.removeUnderscores(title.fragment)
             }
-            .subscribe({
-                if (it.edit?.editSucceeded == true) {
-                    waitForUpdatedRevision(context, wiki, title, it.edit.newRevId, notificationId)
+
+            topic?.let {
+                val submitResponse = ServiceFactory.get(wiki).postEditSubmitSuspend(
+                    title.prefixedText, topic.id.toString(), null,
+                    DIRECT_REPLY_EDIT_COMMENT, if (AccountUtil.isLoggedIn) "user" else null, null, replyText,
+                    talkPageResponse.await().revision, token.await(), null, null
+                )
+                if (submitResponse.edit?.editSucceeded == true) {
+                    waitForUpdatedRevision(context, wiki, title, submitResponse.edit.newRevId, notificationId)
                 } else {
                     fallBackToTalkPage(context, title)
                 }
-            }, {
-                L.e(it)
-                fallBackToTalkPage(context, title)
-            })
+            }
+        }
     }
 
-    private fun waitForUpdatedRevision(context: Context, wiki: WikiSite, title: PageTitle,
-                                       newRevision: Long, notificationId: Int) {
-        ServiceFactory.getRest(wiki).getTalkPage(title.prefixedText)
-            .delay(2, TimeUnit.SECONDS)
-            .subscribeOn(Schedulers.io())
-            .map {
-                if (it.revision < newRevision) {
-                    throw IllegalStateException()
-                }
-                it.revision
-            }
-            .retry(20) {
-                (it is IllegalStateException) || (it is HttpStatusException && it.code == 404)
-            }
-            .observeOn(AndroidSchedulers.mainThread())
-            .doOnTerminate {
-                cancelNotification(context, notificationId)
-            }
-            .subscribe({
-                // revisionForUndo = it
-                Toast.makeText(context, R.string.notifications_direct_reply_success, Toast.LENGTH_LONG).show()
-            }, {
-                L.e(it)
-                fallBackToTalkPage(context, title)
-            })
+    private suspend fun waitForUpdatedRevision(context: Context, wiki: WikiSite, title: PageTitle, newRevision: Long, notificationId: Int) {
+        var talkPageResponse = ServiceFactory.getRest(wiki).getTalkPage(title.prefixedText)
```

Comment:
Can we make it so that `getTalkPage()` is called just once (inside the loop), instead of twice?

Reply:
Yes!